nemo/collections/asr/parts/k2/maploss.py

```diff
@@ -0,0 +1,269 @@
+
+import k2
+import torch
+
+from nemo.collections.asr.parts.k2.autograd import sparse_abs
+from nemo.collections.asr.parts.k2.utils import (
+    create_sparse_wrapped,
+    create_supervision,
+    get_tot_objf_and_finite_mask,
+    load_graph,
+    make_blank_first,
+    prep_padded_densefsavec,
+    shift_labels_inpl,
+)
+from nemo.utils import logging
+
+
+class MAPLoss(torch.nn.Module):
+    """
+    Maximum a Posteriori Probability criterion.
+    It is implemented as Lattice-Free Maximum Mutual Information (LF-MMI) 
+    and LF-boosted-MMI (LF-bMMI) losses.
+    
+    Based on https://github.com/k2-fsa/snowfall/blob/master/snowfall/objectives/mmi.py
+    """
+
+    def __init__(
+        self,
+        num_classes: int,
+        blank: int,
+        reduction: str = "mean",
+        topo_type: str = "default",
+        topo_with_selfloops: bool = True,
+        loss_type: str = "mmi",
+        token_lm: Optional[Union[k2.Fsa, str]] = None,
+        intersect_pruned: bool = False,
+        boost_coeff: float = 0.0,
+        **kwargs,
+    ):
+        super().__init__()
+        self.num_classes = num_classes
+        self.blank = blank
+        self.reduction = reduction
+        self.loss_type = loss_type
+        self.boost_coeff = boost_coeff
+        self.intersect_calc_scores = (
+            self._intersect_calc_scores_mmi_pruned if intersect_pruned else self._intersect_calc_scores_mmi_exact
+        )
+        self.topo_type = topo_type
+        self.topo_with_selfloops = topo_with_selfloops
+        self.pad_fsavec = topo_type == "compact"
+        self.graph_compiler = None
+        if token_lm is None:
+            logging.warning(
+                f"""token_lm is empty. 
+                            Trainable token_lm is not supported yet. 
+                            Please call .update_graph(token_lm) before using."""
+            )
+        else:
+            self.lm_graph = load_graph(token_lm) if isinstance(token_lm, str) else token_lm
+            if self.lm_graph is None:
+                raise ValueError(f"""lm_graph is empty.""")
+            else:
+                self.update_graph(self.lm_graph)
+
+    def update_graph(self, graph: k2.Fsa):
+        self.lm_graph = graph
+        lm_graph = self.lm_graph.clone()
+        if hasattr(lm_graph, "aux_labels"):
+            delattr(lm_graph, "aux_labels")
+        if self.pad_fsavec:
+            shift_labels_inpl([lm_graph], 1)
+        if self.loss_type == "mmi":
+            from nemo.collections.asr.parts.k2.graph_compilers import MmiTrainingGraphCompiler as compiler
+        else:
+            raise ValueError(f"Invalid value of `loss_type`: {self.loss_type}.")
+        self.graph_compiler = compiler(self.num_classes, self.topo_type, self.topo_with_selfloops, aux_graph=lm_graph,)
+
+    def _intersect_calc_scores_mmi_exact(
+        self, dense_fsa_vec: k2.DenseFsaVec, num_graphs: k2.Fsa, den_graph: k2.Fsa, return_lats: bool = True,
```

Comment:
Same as above

Reply:
Done in [7b2b4a4](https://github.com/NVIDIA/NeMo/pull/3351/commits/7b2b4a4ba4e34848d526d8c4ac53903bfbfe2fd2)